### PR TITLE
[expo-notifications][iOS] Swift conversion 2: push token module

### PR DESCRIPTION
**Pull request by douglowder**
# Why

Continuing conversion of expo-notifications iOS modules to Swift

# How

Converted PushTokenModule to Swift. Replaced the complex listener mechanism with a simple ExpoAppDelegateSubscriber class that fires NSNotifications when device tokens are received.

# Test Plan

- Runs successfully on iOS device with test app
- CI should pass

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by douglowder** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/expo/expo/32612?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#32612** <a href="https://app.graphite.dev/github/pr/expo/expo/32612?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#32069** <a href="https://app.graphite.dev/github/pr/expo/expo/32069?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @douglowder and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by vonovak — CHANGES_REQUESTED** [review]
👏👏 👏 Screenshot 2024-11-05 at 12 14 16


Left some stylistic comments, and also I believe there are 2 bugs to be resolved :)

File: packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import MachO
+
+let onDevicePushTokenEventName = "onDevicePushToken"
+
+public class PushTokenModule: Module {
+  var promiseNotYetResolved: Promise?
+
+  public func onNotificationFailure(error: any Error) {
+    if let promiseNotYetResolved = promiseNotYetResolved {
+      promiseNotYetResolved.reject(error)
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift:14]: This function should have been removed, everything is handled in `onNotificationResult` now.

File: packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    }
+  }
+
+  @objc
+  public func onNotificationResult(notification: Notification) {
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift:13]: I'll come up with a better name...

File: packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    OnStopObserving {
+      // swiftlint:disable:next notification_center_detachment
+      NotificationCenter.default.removeObserver(self)
+    }
+
+    AsyncFunction("getDevicePushTokenAsync") { (promise: Promise) in
+      Task { @MainActor in
+        if promiseNotYetResolved != nil {
+          promise.reject("E_AWAIT_PROMISE", "Another async call to this method is in progress. Await the first Promise.")
```
Comment (vonovak) [inline packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift:49]:
```suggestion
          promise.reject("E_AWAIT_PROMISE", "Another async call to 'getDevicePushTokenAsync' method is in progress. Await the first Promise.")
```

**Review by vonovak — CHANGES_REQUESTED** [review]
just one more suggestion :)

**Review by vonovak — APPROVED** [review]

File: packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      self.sendEvent(onDevicePushTokenEventName, ["devicePushToken": deviceToken])
+    }
+  }
+
+  public func definition() -> ModuleDefinition {
+    Name("ExpoPushTokenManager")
+
+    Events([onDevicePushTokenEventName])
+
+    OnStartObserving {
```
Comment (tsapeta) [inline packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift:32]: Let's use the new approach for `OnStartObserving`/`OnStopObserving`, where we provide the event name.
```suggestion
    OnStartObserving(onDevicePushTokenEventName) {
```

File: packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      NotificationCenter.default.addObserver(
+        self,
+        selector: #selector(onExpoNotificationsRegistrationResult),
+        name: PushTokenAppDelegateSubscriber.ExpoNotificationsRegistrationResult,
+        object: nil
+      )
+    }
+
+    OnStopObserving {
```
Comment (tsapeta) [inline packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift:41]:
```suggestion
    OnStopObserving(onDevicePushTokenEventName) {
```

File: packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    AsyncFunction("getDevicePushTokenAsync") { (promise: Promise) in
+      Task { @MainActor in
+        if promiseNotYetResolved != nil {
+          promise.reject("E_AWAIT_PROMISE", "Another async call to this method is in progress. Await the first Promise.")
+        }
+        promiseNotYetResolved = promise
+        UIApplication.shared.registerForRemoteNotifications()
+      }
+    }
```
Comment (tsapeta) [inline packages/expo-notifications/ios/EXNotifications/PushToken/PushTokenModule.swift:54]: If you're using the Promise, switching to the async context makes no sense and becomes much slower. Just use `.runOnQueue(.main)` modifier to run it on the main queue.

```suggestion
    AsyncFunction("getDevicePushTokenAsync") { (promise: Promise) in
      if promiseNotYetResolved != nil {
        promise.reject("E_AWAIT_PROMISE", "Another async call to this method is in progress. Await the first Promise.")
      }
      promiseNotYetResolved = promise
      UIApplication.shared.registerForRemoteNotifications()
    }
    .runOnQueue(.main)
```

**Review by tsapeta — CHANGES_REQUESTED** [review]

**Review by tsapeta — APPROVED** [review]

**Pull request merged**